mantle/kola/harness.go

```diff
@@ -493,12 +496,55 @@ type externalTestMeta struct {
 	Tags          string `json:",tags,omitempty"`
 }
 
-func registerExternalTest(testname, executable, dependencydir, ignition string, meta externalTestMeta) error {
+// metadataFromTestBinary extracts JSON-in-comment like:
+// #!/bin/bash
+// # kola: { "tags": ["ignition"], "architectures": ["x86_64"] }
```

Comment:
Something about this offends lessons learned from a previous life: comments should be just logic.

But I also don't see a quick/easy way to what you're after without implementing a proper API. I'd almost rather the test imbedded in the JSON. That said, I think I'm okay with this. But I'd like to have someone else weigh in.

Reply:
Clearly we already have a lot of shell-in-YAML when the metadata is larger than the shell; the postprocess scripts in rpm-ostree manifests, various Kubernetes pods, etc.
(Hopefully everyone prefers YAML over JSON for this type of stuff)

In this case, the metadata is tiny relative to the code which I think argues for this approach.

I don't think using magic comments is unusual though in this situation; in fact, `#!/bin/sh` is probably *the original* magic comment that tells the kernel how to run the code!

https://github.com/DanielKeep/cargo-script#scripts is another example of this.

Reply:
Your points are valid @cgwalters, and clearly this is a pattern that has been done in other code-bases. Like I said, I can see the "why" and my "objection" is one of personal taste. So in the sake of moving this forward, LGTM.

Reply:
To be clear I think this *definitely* worth debating/bikeshedding because we're talking about an "API" that will end up being used by potentially a bunch of our other repos, and will be hard to change later.  There may be better alternatives still!  But, if it's this or "shell-in-JSON/YAML" I lean a bit towards this given how small the metadata is.

One other thing I was thinking about is taking a page from systemd generators; basically we support `/usr/lib/coreos-assembler/tests/kola-generators/` which is a list of binaries that we run, and they output JSON or something that we parse.

That would solve the "ELF binaries can't easily add `# kola:` in the first 4k" problem and others.

But it would require a bit more infrastructure test side.

Reply:
> There may be better alternatives still! But, if it's this or "shell-in-JSON/YAML" I lean a bit towards this given how small the metadata is.

I agree that inline better for readability, at least with the current implementation. My honest gut reaction when I saw this PR was to require external binaries to have a `meta` and `run` CLI,  such that Kola could get the JSON in a programmatic manner. I resisted the urge to bike-shed on the design since that would impose a higher cost to develop tests; this is working code.

With a CLI "API" the advantage could be pretty much any executable regardless of the comment style. Going a step further, the `meta` could then provide requirements by returning something like:
```
{
   "test_api": "1.0",
   "min_version": "46.82",
   "tags": ["ignition"], 
   "architectures": ["x86_64"],
   "hardware":  ["TPM"]
 }
```

Then we harden for the future by having a versioned API. 

But my hesitancy in suggesting that is that I'm not sure that cost is worth that. The tests are tightly coupled to our current feature set, and we change things so rapidly and I hadn't fully fleshed out my thinking on it.

Reply:
I like the CLI API idea in principle, though one concern is that it's not easy from the outside to tell whether the test script/binary implements it or whether running it might actually end up running the test and e.g. reboot your computer. IOW, I think it'd have to be made mandatory rather than optional?

One thing I like about the external host test model is that tests can literally be [two lines long](https://github.com/coreos/fedora-coreos-config/blob/testing-devel/tests/kola/basic). Having to handle commands would be unfortunate.

I understand the hesitation though re. inlining as a comment. It shouldn't be the only way, though as a shortcut for not having a separate JSON file? WFM.

Reply:
> I understand the hesitation though re. inlining as a comment. It shouldn't be the only way, though as a shortcut for not having a separate JSON file? WFM.

:100:  I hesitated to make the suggestion because of the burden to implement a test. But once we start talking API or making explicit promises, that when my spidey sense tingles. 

As long as we are the sole consumers, I'm okay with YOLO'ing this and when we have interest later we can revisit. For that matter, we can make this "version 1.0" of the "API." 

Perhaps a clear path to balance future needs against the immediate friction is that we put a commit comment along the lines of "Yeah, we are totally going to change this later, so ask first."